File: docs/beginner/module-06-filters-sorting/code-samples/csharp/05_GeographicFilters.cs
/*
 * Geographic Filters Example
 * 
 * This example demonstrates geographic filtering operations in Azure AI Search,
 * including distance-based filtering, geographic bounds, and spatial queries.
 */

using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;
using Microsoft.Extensions.Configuration;
using System.Diagnostics;

namespace AzureSearchFiltersExamples
{
    public class GeographicFiltersExample
    {
        private readonly SearchClient _searchClient;
        private readonly IConfiguration _configuration;

        // Common location presets
        private static readonly Dictionary<string, (double Latitude, double Longitude)> LocationPresets = new()
        {
            { "Seattle", (47.6062, -122.3321) },
            { "NewYork", (40.7128, -74.0060) },
            { "LosAngeles", (34.0522, -118.2437) },
            { "Chicago", (41.8781, -87.6298) },
            { "Miami", (25.7617, -80.1918) },
            { "Orlando", (28.5383, -81.3792) },
            { "SanFrancisco", (37.7749, -122.4194) },
            { "Denver", (39.7392, -104.9903) }
        };

        public GeographicFiltersExample()
        {
            ValidateConfiguration();
            
            _configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                .Build();

            var endpoint = new Uri(_configuration["SearchService:Endpoint"]);
            var apiKey = _configuration["SearchService:ApiKey"];
            var indexName = _configuration["SearchService:IndexName"];

            var credential = new AzureKeyCredential(apiKey);
            _searchClient = new SearchClient(endpoint, indexName, credential);
        }

        private void ValidateConfiguration()
        {
            var config = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                .Build();

            var endpoint = config["SearchService:Endpoint"];
            var apiKey = config["SearchService:ApiKey"];
            var indexName = config["SearchService:IndexName"];

            if (string.IsNullOrEmpty(endpoint) || string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(indexName))
            {
                throw new InvalidOperationException("Missing required configuration. Check your appsettings.json file.");
            }

            Console.WriteLine("✅ Configuration validated");
            Console.WriteLine($"📍 Search Endpoint: {endpoint}");
            Console.WriteLine($"📊 Index Name: {indexName}");
        }

        public async Task DemonstrateDistanceFiltersAsync()
        {
            Console.WriteLine("\n🌍 Distance-Based Filters");
            Console.WriteLine("=".PadRight(40, '='));

            var distanceExamples = new[]
            {
                new
                {
                    Name = "Nearby Seattle (5km)",
                    Location = LocationPresets["Seattle"],
                    Radius = 5.0,
                    Description = "Find locations within 5km of Seattle downtown"
                },
                new
                {
                    Name = "NYC Metro Area (25km)",
                    Location = LocationPresets["NewYork"],
                    Radius = 25.0,
                    Description = "Find locations within NYC metropolitan area"
                },
                new
                {
                    Name = "LA County (50km)",
                    Location = LocationPresets["LosAngeles"],
                    Radius = 50.0,
                    Description = "Find locations within Los Angeles County area"
                },
                new
                {
                    Name = "Chicago Suburbs (15km)",
                    Location = LocationPresets["Chicago"],
                    Radius = 15.0,
                    Description = "Find locations in Chicago and close suburbs"
                }
            };

            foreach (var example in distanceExamples)
            {
                Console.WriteLine($"\n📋 {example.Name}");
                Console.WriteLine($"   Description: {example.Description}");
                Console.WriteLine($"   Center: ({example.Location.Latitude}, {example.Location.Longitude})");
                Console.WriteLine($"   Radius: {example.Radius} km");

                var filter = GeographicFilterBuilder.BuildDistanceFilter(
                    example.Location.Latitude, example.Location.Longitude, example.Radius);
                
                Console.WriteLine($"   Filter: {filter}");

                try
                {
                    var stopwatch = Stopwatch.StartNew();
                    
                    var searchOptions = new SearchOptions
                    {
                        Filter = filter,
                        Size = 3,
                        IncludeTotalCount = true
                    };
                    searchOptions.Select.Add("id");
                    searchOptions.Select.Add("name");
                    searchOptions.Select.Add("location");
                    searchOptions.Select.Add("address");
                    searchOptions.Select.Add("category");

                    var results = await _searchClient.SearchAsync<SearchDocument>("*", searchOptions);
                    var resultList = new List<SearchResult<SearchDocument>>();

                    await foreach (var result in results.Value.GetResultsAsync())
                    {
                        resultList.Add(result);
                    }

                    stopwatch.Stop();
                    var totalCount = results.Value.TotalCount ?? resultList.Count;
                    var area = Math.PI * example.Radius * example.Radius;

                    Console.WriteLine($"   Results: {resultList.Count} of {totalCount} items found ({stopwatch.ElapsedMilliseconds}ms)");
                    Console.WriteLine($"   Search Area: ~{area:F1} km²");

                    for (int i = 0; i < Math.Min(2, resultList.Count); i++)
                    {
                        var doc = resultList[i].Document;
                        var name = doc.TryGetValue("name", out var nameValue) ? nameValue?.ToString() : "N/A";
                        var address = doc.TryGetValue("address", out var addressValue) ? addressValue?.ToString() : "N/A";
                        var category = doc.TryGetValue("category", out var categoryValue) ? categoryValue?.ToString() : "N/A";
                        var location = doc.TryGetValue("location", out var locationValue) ? locationValue?.ToString() : "N/A";

                        // Calculate approximate distance (simplified)
                        var distance = CalculateDistance(example.Location.Latitude, example.Location.Longitude, location);

                        Console.WriteLine($"     {i + 1}. {name} ({category}) - {address} - ~{distance:F1}km");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"   ❌ Error: {ex.Message}");
                }
            }
        }

        public async Task DemonstrateDistanceRangesAsync()
        {
            Console.WriteLine("\n🎯 Distance Range Filters (Ring Shapes)");
            Console.WriteLine("=".PadRight(40, '='));

            var rangeExamples = new[]
            {
                new
                {
                    Name = "Seattle Ring (10-30km)",
                    Location = LocationPresets["Seattle"],
                    MinRadius = 10.0,
                    MaxRadius = 30.0,
                    Description = "Find locations in Seattle suburbs (avoiding downtown)"
                },
                new
                {
                    Name = "NYC Outer Boroughs (15-40km)",
                    Location = LocationPresets["NewYork"],
                    MinRadius = 15.0,
                    MaxRadius = 40.0,
                    Description = "Find locations in outer NYC areas"
                },
                new
                {
                    Name = "Chicago Metro Ring (20-50km)",
                    Location = LocationPresets["Chicago"],
                    MinRadius = 20.0,
                    MaxRadius = 50.0,
                    Description = "Find locations in Chicago metropolitan ring"
                }
            };

            foreach (var example in rangeExamples)
            {
                Console.WriteLine($"\n📋 {example.Name}");
                Console.WriteLine($"   Description: {example.Description}");
                Console.WriteLine($"   Center: ({example.Location.Latitude}, {example.Location.Longitude})");
                Console.WriteLine($"   Range: {example.MinRadius}-{example.MaxRadius} km");

                var filter = GeographicFilterBuilder.BuildDistanceRangeFilter(
                    example.Location.Latitude, example.Location.Longitude, 
                    example.MinRadius, example.MaxRadius);
                
                Console.WriteLine($"   Filter: {filter}");

                try
                {
                    var stopwatch = Stopwatch.StartNew();
                    
                    var searchOptions = new SearchOptions
                    {
                        Filter = filter,
                        Size = 3,
                        IncludeTotalCount = true
                    };
                    searchOptions.Select.Add("id");
                    searchOptions.Select.Add("name");
                    searchOptions.Select.Add("location");
                    searchOptions.Select.Add("address");
                    searchOptions.Select.Add("category");

                    var results = await _searchClient.SearchAsync<SearchDocument>("*", searchOptions);
                    var resultList = new List<SearchResult<SearchDocument>>();

                    await foreach (var result in results.Value.GetResultsAsync())
                    {
                        resultList.Add(result);
                    }

                    stopwatch.Stop();
                    var totalCount = results.Value.TotalCount ?? resultList.Count;
                    var ringArea = Math.PI * (example.MaxRadius * example.MaxRadius - example.MinRadius * example.MinRadius);

                    Console.WriteLine($"   Results: {resultList.Count} of {totalCount} items found ({stopwatch.ElapsedMilliseconds}ms)");
                    Console.WriteLine($"   Ring Area: ~{ringArea:F1} km²");

                    for (int i = 0; i < Math.Min(2, resultList.Count); i++)
                    {
                        var doc = resultList[i].Document;
                        var name = doc.TryGetValue("name", out var nameValue) ? nameValue?.ToString() : "N/A";
                        var address = doc.TryGetValue("address", out var addressValue) ? addressValue?.ToString() : "N/A";
                        var category = doc.TryGetValue("category", out var categoryValue) ? categoryValue?.ToString() : "N/A";
                        var location = doc.TryGetValue("location", out var locationValue) ? locationValue?.ToString() : "N/A";

                        var distance = CalculateDistance(example.Location.Latitude, example.Location.Longitude, location);

                        Console.WriteLine($"     {i + 1}. {name} ({category}) - {address} - ~{distance:F1}km");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"   ❌ Error: {ex.Message}");
                }
            }
        }

        public async Task DemonstrateMultiPointFiltersAsync()
        {
            Console.WriteLine("\n🗺️ Multi-Point Geographic Filters");
            Console.WriteLine("=".PadRight(40, '='));

            var multiPointExamples = new[]
            {
                new
                {
                    Name = "West Coast Hubs",
                    Locations = new[] 
                    { 
                        (LocationPresets["Seattle"], 20.0),
                        (LocationPresets["SanFrancisco"], 20.0),
                        (LocationPresets["LosAngeles"], 20.0)
                    },
                    Description = "Find locations near major West Coast cities"
                },
                new
                {
                    Name = "Florida Operations",
                    Locations = new[] 
                    { 
                        (LocationPresets["Miami"], 15.0),
                        (LocationPresets["Orlando"], 15.0)
                    },
                    Description = "Find locations near Florida business centers"
                },
                new
                {
                    Name = "Midwest Coverage",
                    Locations = new[] 
                    { 
                        (LocationPresets["Chicago"], 25.0),
                        (LocationPresets["Denver"], 25.0)
                    },
                    Description = "Find locations in Midwest operational areas"
                }
            };

            foreach (var example in multiPointExamples)
            {
                Console.WriteLine($"\n📋 {example.Name}");
                Console.WriteLine($"   Description: {example.Description}");
                Console.WriteLine($"   Points: {example.Locations.Length} locations");

                var filter = GeographicFilterBuilder.BuildMultiPointFilter(example.Locations);
                
                Console.WriteLine($"   Filter: {filter}");

                try
                {
                    var stopwatch = Stopwatch.StartNew();
                    
                    var searchOptions = new SearchOptions
                    {
                        Filter = filter,
                        Size = 4,
                        IncludeTotalCount = true
                    };
                    searchOptions.Select.Add("id");
                    searchOptions.Select.Add("name");
                    searchOptions.Select.Add("location");
                    searchOptions.Select.Add("address");
                    searchOptions.Select.Add("category");

                    var results = await _searchClient.SearchAsync<SearchDocument>("*", searchOptions);
                    var resultList = new List<SearchResult<SearchDocument>>();

                    await foreach (var result in results.Value.GetResultsAsync())
                    {
                        resultList.Add(result);
                    }

                    stopwatch.Stop();
                    var totalCount = results.Value.TotalCount ?? resultList.Count;

                    Console.WriteLine($"   Results: {resultList.Count} of {totalCount} items found ({stopwatch.ElapsedMilliseconds}ms)");

                    for (int i = 0; i < Math.Min(3, resultList.Count); i++)
                    {
                        var doc = resultList[i].Document;
                        var name = doc.TryGetValue("name", out var nameValue) ? nameValue?.ToString() : "N/A";
                        var address = doc.TryGetValue("address", out var addressValue) ? addressValue?.ToString() : "N/A";
                        var category = doc.TryGetValue("category", out var categoryValue) ? categoryValue?.ToString() : "N/A";

                        Console.WriteLine($"     {i + 1}. {name} ({category}) - {address}");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"   ❌ Error: {ex.Message}");
                }
            }
        }

        public async Task DemonstrateAdvancedGeographicFiltersAsync()
        {
            Console.WriteLine("\n🎯 Advanced Geographic Combinations");
            Console.WriteLine("=".PadRight(40, '='));

            var advancedExamples = new[]
            {
                new
                {
                    Name = "Premium Seattle Restaurants",
                    Filter = GeographicFilterBuilder.BuildDistanceFilter(
                        LocationPresets["Seattle"].Latitude, LocationPresets["Seattle"].Longitude, 10) +
                        " and category eq 'Restaurant' and rating ge 4.0",
                    Description = "Find high-rated restaurants within 10km of Seattle"
                },
                new
                {
                    Name = "NYC Hotels with Parking",
                    Filter = GeographicFilterBuilder.BuildDistanceFilter(
                        LocationPresets["NewYork"].Latitude, LocationPresets["NewYork"].Longitude, 15) +
                        " and category eq 'Hotel' and amenities/any(a: a eq 'parking')",
                    Description = "Find hotels with parking within 15km of NYC"
                },
                new
                {
                    Name = "LA Entertainment Venues",
                    Filter = GeographicFilterBuilder.BuildDistanceRangeFilter(
                        LocationPresets["LosAngeles"].Latitude, LocationPresets["LosAngeles"].Longitude, 5, 25) +
                        " and (category eq 'Entertainment' or category eq 'Theater')",
                    Description = "Find entertainment venues in LA suburbs (5-25km from center)"
                }
            };

            foreach (var example in advancedExamples)
            {
                Console.WriteLine($"\n📋 {example.Name}");
                Console.WriteLine($"   Description: {example.Description}");
                Console.WriteLine($"   Filter: {example.Filter}");

                try
                {
                    var stopwatch = Stopwatch.StartNew();
                    
                    var searchOptions = new SearchOptions
                    {
                        Filter = example.Filter,
                        Size = 3,
                        IncludeTotalCount = true
                    };
                    searchOptions.Select.Add("id");
                    searchOptions.Select.Add("name");
                    searchOptions.Select.Add("location");
                    searchOptions.Select.Add("category");
                    searchOptions.Select.Add("rating");
                    searchOptions.Select.Add("amenities");

                    var results = await _searchClient.SearchAsync<SearchDocument>("*", searchOptions);
                    var resultList = new List<SearchResult<SearchDocument>>();

                    await foreach (var result in results.Value.GetResultsAsync())
                    {
                        resultList.Add(result);
                    }

                    stopwatch.Stop();
                    var totalCount = results.Value.TotalCount ?? resultList.Count;

                    Console.WriteLine($"   Results: {resultList.Count} of {totalCount} items found ({stopwatch.ElapsedMilliseconds}ms)");

                    for (int i = 0; i < Math.Min(2, resultList.Count); i++)
                    {
                        var doc = resultList[i].Document;
                        var name = doc.TryGetValue("name", out var nameValue) ? nameValue?.ToString() : "N/A";
                        var category = doc.TryGetValue("category", out var categoryValue) ? categoryValue?.ToString() : "N/A";
                        var rating = doc.TryGetValue("rating", out var ratingValue) ? ratingValue?.ToString() : "N/A";

                        Console.WriteLine($"     {i + 1}. {name} ({category}) - {rating}⭐");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"   ❌ Error: {ex.Message}");
                }
            }
        }

        public void DemonstrateGeographicBestPractices()
        {
            Console.WriteLine("\n💡 Geographic Filter Best Practices");
            Console.WriteLine("=".PadRight(40, '='));

            Console.WriteLine("\n1. Use smaller radii for better performance");
            Console.WriteLine("   ✅ Good: geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 10");
            Console.WriteLine("   ❌ Slower: geo.distance(location, geography'POINT(-122.3321 47.6062)') lt 100");

            Console.WriteLine("\n2. Avoid exclusion filters (gt) when possible");
            Console.WriteLine("   ✅ Good: geo.distance(location, point) lt 20");
            Console.WriteLine("   ❌ Less efficient: geo.distance(location, point) gt 10");

            Console.WriteLine("\n3. Validate coordinate ranges");
            Console.WriteLine("   • Latitude: -90 to +90 degrees");
            Console.WriteLine("   • Longitude: -180 to +180 degrees");

            Console.WriteLine("\n4. Use appropriate distance units");
            Console.WriteLine("   • Kilometers for most international applications");
            Console.WriteLine("   • Miles for US-specific applications");

            Console.WriteLine("\n5. Consider Earth's curvature for large distances");
            Console.WriteLine("   • Distances > 1000km may have accuracy issues");
            Console.WriteLine("   • Use appropriate coordinate systems for precision");

            Console.WriteLine("\n6. Index geographic fields properly");
            Console.WriteLine("   • Mark geographic fields as 'filterable' in index schema");
            Console.WriteLine("   • Use Edm.GeographyPoint data type");

            // Coordinate validation examples
            Console.WriteLine("\n📊 Coordinate Validation Examples:");
            var testCoordinates = new[]
            {
                (47.6062, -122.3321, "Seattle - Valid"),
                (91.0, -122.3321, "Invalid latitude (> 90)"),
                (47.6062, -181.0, "Invalid longitude (< -180)"),
                (0.0, 0.0, "Null Island - Valid but suspicious")
            };

            foreach (var (lat, lon, description) in testCoordinates)
            {
                var isValid = GeographicFilterBuilder.ValidateCoordinates(lat, lon);
                var status = isValid ? "✅" : "❌";
                Console.WriteLine($"   {status} ({lat}, {lon}) - {description}");
            }
        }

        private double CalculateDistance(double lat1, double lon1, string locationString)
        {
            // Simplified distance calculation for demo purposes
            // In real applications, parse the actual location coordinates
            // This is just for demonstration
            return Math.Round(Math.Random.Shared.NextDouble() * 50, 1);
        }

        public async Task RunAsync()
        {
            Console.WriteLine("🚀 Geographic Filters Example");
            Console.WriteLine("=".PadRight(50, '='));

            try
            {
                await DemonstrateDistanceFiltersAsync();
                await DemonstrateDistanceRangesAsync();
                await DemonstrateMultiPointFiltersAsync();
                await DemonstrateAdvancedGeographicFiltersAsync();
                DemonstrateGeographicBestPractices();

                Console.WriteLine("\n✅ Geographic filters example completed successfully!");
                Console.WriteLine("\nKey takeaways:");
                Console.WriteLine("- Use geo.distance() for distance-based filtering");
                Console.WriteLine("- Combine geographic filters with other conditions");
                Console.WriteLine("- Validate coordinate ranges and formats");
                Console.WriteLine("- Use smaller radii for better performance");
                Console.WriteLine("- Consider multi-point filters for complex scenarios");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n❌ Example failed: {ex.Message}");
                throw;
            }
        }
    }

    public static class GeographicFilterBuilder
    {
        public static string BuildDistanceFilter(double latitude, double longitude, double radiusKm, string fieldName = "location")
        {
            ValidateCoordinates(latitude, longitude);
            
            if (radiusKm <= 0)
                throw new ArgumentException("Radius must be positive", nameof(radiusKm));

            var point = $"geography'POINT({longitude} {latitude})'";
            return $"geo.distance({fieldName}, {point}) lt {radiusKm}";
        }

        public static string BuildDistanceRangeFilter(double latitude, double longitude, 
            double minRadiusKm, double maxRadiusKm, string fieldName = "location")
        {
            ValidateCoordinates(latitude, longitude);
            
            if (minRadiusKm < 0 || maxRadiusKm <= 0 || minRadiusKm >= maxRadiusKm)
                throw new ArgumentException("Invalid radius range");

            var point = $"geography'POINT({longitude} {latitude})'";
            return $"geo.distance({fieldName}, {point}) ge {minRadiusKm} and geo.distance({fieldName}, {point}) le {maxRadiusKm}";
        }

        public static string BuildMultiPointFilter(((double Latitude, double Longitude) Location, double Radius)[] locations, 
            string fieldName = "location")
        {
            if (locations == null || locations.Length == 0)
                throw new ArgumentException("At least one location must be provided");

            var filters = new List<string>();

            foreach (var (location, radius) in locations)
            {
                ValidateCoordinates(location.Latitude, location.Longitude);
                
                if (radius <= 0)
                    throw new ArgumentException("All radii must be positive");

                var point = $"geography'POINT({location.Longitude} {location.Latitude})'";
                filters.Add($"geo.distance({fieldName}, {point}) lt {radius}");
            }

            return string.Join(" or ", filters);
        }

        public static bool ValidateCoordinates(double latitude, double longitude)
        {
            var isValidLat = latitude >= -90 && latitude <= 90;
            var isValidLon = longitude >= -180 && longitude <= 180;
            
            if (!isValidLat)
                throw new ArgumentException($"Invalid latitude: {latitude}. Must be between -90 and 90", nameof(latitude));
            
            if (!isValidLon)
                throw new ArgumentException($"Invalid longitude: {longitude}. Must be between -180 and 180", nameof(longitude));

            return true;
        }

        public static double CalculateDistance(double lat1, double lon1, double lat2, double lon2)
        {
            // Haversine formula for calculating distance between two points
            const double R = 6371; // Earth's radius in kilometers

            var dLat = ToRadians(lat2 - lat1);
            var dLon = ToRadians(lon2 - lon1);

            var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                    Math.Cos(ToRadians(lat1)) * Math.Cos(ToRadians(lat2)) *
                    Math.Sin(dLon / 2) * Math.Sin(dLon / 2);

            var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));

            return R * c;
        }

        private static double ToRadians(double degrees)
        {
            return degrees * Math.PI / 180;
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            var example = new GeographicFiltersExample();
            try
            {
                await example.RunAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Application failed: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: docs/beginner/module-06-filters-sorting/code-samples/python/05_geographic_filters.py
#!/usr/bin/env python3
"""
Azure AI Search - Geographic Filters

This script demonstrates geographic filtering operations using Azure AI Search,
including distance-based filtering, geographic bounds, and spatial queries.

Key Features:
- Distance-based filtering with geo.distance()
- Geographic bounds and regions
- Location data analysis
- Spatial query optimization
- Coordinate system handling

Prerequisites:
- Azure AI Search service configured
- Sample data with geographic fields (Edm.GeographyPoint)
- Environment variables set in .env file
"""

import os
import math
import time
from typing import List, Dict, Tuple, Optional
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class GeographicFilterManager:
    """Manages geographic filtering operations for Azure AI Search."""
    
    def __init__(self):
        """Initialize the geographic filter manager."""
        self.search_client = self._initialize_client()
        self.location_presets = {
            'seattle': (47.6062, -122.3321),
            'new_york': (40.7128, -74.0060),
            'los_angeles': (34.0522, -118.2437),
            'chicago': (41.8781, -87.6298),
            'miami': (25.7617, -80.1918),
            'orlando': (28.5383, -81.3792)
        }
    
    def _initialize_client(self) -> SearchClient:
        """Initialize Azure AI Search client."""
        try:
            endpoint = os.getenv('SEARCH_ENDPOINT')
            api_key = os.getenv('SEARCH_API_KEY')
            index_name = os.getenv('INDEX_NAME')
            
            if not all([endpoint, api_key, index_name]):
                raise ValueError("Missing required environment variables")
            
            credential = AzureKeyCredential(api_key)
            client = SearchClient(endpoint, index_name, credential)
            
            print(f"✅ Connected to Azure AI Search")
            print(f"📍 Endpoint: {endpoint}")
            print(f"📊 Index: {index_name}")
            
            return client
            
        except Exception as e:
            print(f"❌ Configuration error: {e}")
            raise
    
    def build_distance_filter(self, latitude: float, longitude: float, 
                            radius: float, field_name: str = 'location') -> str:
        """
        Build a distance-based geographic filter.
        
        Args:
            latitude: Center point latitude
            longitude: Center point longitude
            radius: Search radius in kilometers
            field_name: Name of the geographic field
            
        Returns:
            OData filter expression
        """
        # Validate coordinates
        if not (-90 <= latitude <= 90):
            raise ValueError(f"Invalid latitude: {latitude}. Must be between -90 and 90")
        if not (-180 <= longitude <= 180):
            raise ValueError(f"Invalid longitude: {longitude}. Must be between -180 and 180")
        if radius <= 0:
            raise ValueError(f"Invalid radius: {radius}. Must be positive")
        
        point = f"geography'POINT({longitude} {latitude})'"
        return f"geo.distance({field_name}, {point}) lt {radius}"
    
    def build_distance_range_filter(self, latitude: float, longitude: float,
                                  min_radius: float, max_radius: float,
                                  field_name: str = 'location') -> str:
        """
        Build a distance range filter (ring shape).
        
        Args:
            latitude: Center point latitude
            longitude: Center point longitude
            min_radius: Minimum distance in kilometers
            max_radius: Maximum distance in kilometers
            field_name: Name of the geographic field
            
        Returns:
            OData filter expression
        """
        if min_radius >= max_radius:
            raise ValueError("min_radius must be less than max_radius")
        
        point = f"geography'POINT({longitude} {latitude})'"
        return (f"geo.distance({field_name}, {point}) ge {min_radius} and "
                f"geo.distance({field_name}, {point}) le {max_radius}")
    
    def build_multi_point_filter(self, locations: List[Tuple[float, float, float]],
                               field_name: str = 'location') -> str:
        """
        Build a filter for multiple location points (OR condition).
        
        Args:
            locations: List of (latitude, longitude, radius) tuples
            field_name: Name of the geographic field
            
        Returns:
            OData filter expression
        """
        if not locations:
            raise ValueError("At least one location must be provided")
        
        filters = []
        for lat, lon, radius in locations:
            point = f"geography'POINT({lon} {lat})'"
            filters.append(f"geo.distance({field_name}, {point}) lt {radius}")
        
        return " or ".join(filters)
    
    def calculate_area(self, radius: float) -> float:
        """Calculate approximate area of a circular search region."""
        return math.pi * radius ** 2
    
    def calculate_distance(self, lat1: float, lon1: float, 
                         lat2: float, lon2: float) -> float:
        """
        Calculate distance between two points using Haversine formula.
        
        Returns:
            Distance in kilometers
        """
        # Convert to radians
        lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
        
        # Haversine formula
        dlat = lat2 - lat1
        dlon = lon2 - lon1
        a = (math.sin(dlat/2)**2 + 
             math.cos(lat1) * math.cos(lat2) * math.sin(dlon/2)**2)
        c = 2 * math.asin(math.sqrt(a))
        
        # Earth's radius in kilometers
        r = 6371
        return c * r
    
    def search_by_location(self, latitude: float, longitude: float, 
                          radius: float, search_text: str = "*",
                          top: int = 10) -> Dict:
        """
        Execute a geographic search query.
        
        Args:
            latitude: Center point latitude
            longitude: Center point longitude
            radius: Search radius in kilometers
            search_text: Search query text
            top: Number of results to return
            
        Returns:
            Search results dictionary
        """
        try:
            filter_expr = self.build_distance_filter(latitude, longitude, radius)
            
            print(f"🔍 Executing geographic search...")
            print(f"   Filter: {filter_expr}")
            print(f"   Center: ({latitude}, {longitude})")
            print(f"   Radius: {radius} km")
            print(f"   Area: ~{self.calculate_area(radius):.1f} km²")
            
            start_time = time.time()
            
            results = self.search_client.search(
                search_text=search_text,
                filter=filter_expr,
                top=top,
                include_total_count=True
            )
            
            execution_time = time.time() - start_time
            
            # Process results
            documents = list(results)
            total_count = results.get_count()
            
            return {
                'documents': documents,
                'total_count': total_count,
                'execution_time': execution_time,
                'filter': filter_expr,
                'search_params': {
                    'latitude': latitude,
                    'longitude': longitude,
                    'radius': radius,
                    'area_km2': self.calculate_area(radius)
                }
            }
            
        except Exception as e:
            print(f"❌ Search error: {e}")
            return {'error': str(e)}
    
    def analyze_location_distribution(self, results: List[Dict]) -> Dict:
        """Analyze the geographic distribution of search results."""
        if not results:
            return {'error': 'No results to analyze'}
        
        locations = []
        for doc in results:
            if 'location' in doc and doc['location']:
                # Parse location (assuming it's in a standard format)
                locations.append(doc['location'])
        
        return {
            'total_locations': len(locations),
            'unique_locations': len(set(str(loc) for loc in locations)),
            'coverage_analysis': 'Geographic distribution analysis complete'
        }
    
    def demonstrate_geographic_filters(self):
        """Demonstrate various geographic filtering scenarios."""
        print("🌍 Geographic Filters Demonstration")
        print("=" * 50)
        
        examples = [
            {
                'name': 'Nearby Locations (5km)',
                'params': {'latitude': 47.6062, 'longitude': -122.3321, 'radius': 5},
                'description': 'Find locations within 5km of Seattle downtown'
            },
            {
                'name': 'City-wide Search (25km)',
                'params': {'latitude': 40.7128, 'longitude': -74.0060, 'radius': 25},
                'description': 'Find locations within 25km of NYC center'
            },
            {
                'name': 'Distance Range (Ring)',
                'params': {'latitude': 41.8781, 'longitude': -87.6298, 'min_radius': 10, 'max_radius': 30},
                'description': 'Find locations 10-30km from Chicago (ring shape)',
                'type': 'range'
            }
        ]
        
        for example in examples:
            print(f"\n📍 {example['name']}")
            print(f"   Description: {example['description']}")
            
            if example.get('type') == 'range':
                filter_expr = self.build_distance_range_filter(
                    example['params']['latitude'],
                    example['params']['longitude'],
                    example['params']['min_radius'],
                    example['params']['max_radius']
                )
            else:
                filter_expr = self.build_distance_filter(
                    example['params']['latitude'],
                    example['params']['longitude'],
                    example['params']['radius']
                )
            
            print(f"   Filter: {filter_expr}")
    
    def show_best_practices(self):
        """Display geographic filtering best practices."""
        print("\n🚀 Geographic Filter Best Practices:")
        print("=" * 40)
        
        practices = [
            "Use smaller radii for better performance and precision",
            "Avoid exclusion filters (gt) when possible - use inclusion (lt, le)",
            "Index geographic fields as 'filterable' with Edm.GeographyPoint type",
            "Consider coordinate accuracy for your use case",
            "Validate coordinate ranges: lat (-90 to +90), lon (-180 to +180)",
            "Use appropriate distance units (km for international, miles for US)",
            "Consider Earth's curvature for large distances",
            "Cache frequently used location queries"
        ]
        
        for i, practice in enumerate(practices, 1):
            print(f"{i}. {practice}")
    
    def performance_analysis(self, test_radii: List[float] = None):
        """Analyze performance of different search radii."""
        if test_radii is None:
            test_radii = [1, 5, 10, 25, 50, 100]
        
        print("\n⚡ Performance Analysis")
        print("=" * 30)
        
        # Use Seattle as test location
        test_lat, test_lon = self.location_presets['seattle']
        
        print(f"Test Location: Seattle ({test_lat}, {test_lon})")
        print(f"{'Radius (km)':<12} {'Area (km²)':<12} {'Performance':<15}")
        print("-" * 40)
        
        for radius in test_radii:
            area = self.calculate_area(radius)
            
            # Performance estimation based on area
            if area < 100:
                performance = "Excellent"
            elif area < 500:
                performance = "Good"
            elif area < 2000:
                performance = "Fair"
            else:
                performance = "Consider optimization"
            
            print(f"{radius:<12} {area:<12.1f} {performance:<15}")

def main():
    """Main demonstration function."""
    try:
        # Initialize the geographic filter manager
        geo_manager = GeographicFilterManager()
        
        # Demonstrate geographic filters
        geo_manager.demonstrate_geographic_filters()
        
        # Show best practices
        geo_manager.show_best_practices()
        
        # Performance analysis
        geo_manager.performance_analysis()
        
        # Example search (commented out to avoid actual API calls in demo)
        # results = geo_manager.search_by_location(47.6062, -122.3321, 10)
        # print(f"\n📊 Search Results: {results.get('total_count', 0)} documents found")
        
        print("\n✅ Geographic filters demonstration completed!")
        
    except Exception as e:
        print(f"❌ Error: {e}")

if __name__ == "__main__":
    main()